HW - 21.12.07 - Created migration files

Daniel-Cretu opened the pull request.

egorbwork commented on config/app.php:
```diff
@@ -67,7 +67,8 @@
     */
 
-    'timezone' => 'UTC',
+//    'timezone' => 'UTC',
```
I do not recommend to comment this kind of code.

Daniel-Cretu replied:
Understood

egorbwork commented on database/migrations/2021_12_12_170926_alter_products_table.php:
```diff
@@ -0,0 +1,34 @@
+    {
+        Schema::table('products', function (Blueprint $table){
+           $table->string('name',40);
+           $table->decimal('price',10,2)->nullable(true);
```
interesting that now you don't need `after` in adding column.

Daniel-Cretu replied:
Should I change it up somehow to be needed to use 'after' ?

egorbwork replied:
nope it is ok in this way too.

egorbwork commented on database/migrations/2021_12_12_171823_create_warnings_table.php:
```diff
@@ -0,0 +1,32 @@
+    {
+        Schema::create('warnings', function (Blueprint $table) {
+            $table->id();
+            $table->string('name',40);
```
unique is missing

Daniel-Cretu replied:
Added unique to name column by adding an alter migration.

egorbwork replied:
ok, it is ok to be migration for it. But it is still is not change of column, it is creation of unique index.

egorbwork commented on database/migrations/2021_12_12_171955_create_product_warning_table.php:
```diff
@@ -0,0 +1,39 @@
+                ->references('id')->on('warnings')
+                ->cascadeOnUpdate()
+                ->cascadeOnDelete();
+            $table->unique(['product_id', 'warning_id']);
```
instead primary key could be used. But it is not mandatory.

Daniel-Cretu replied:
The main idea is to not be able to insert product 'A' with warning 'B' multiple times.

egorbwork commented on database/migrations/2021_12_12_172732_create_statuses_table.php:
```diff
@@ -0,0 +1,31 @@
+    {
+        Schema::create('statuses', function (Blueprint $table) {
+            $table->id();
+            $table->string('name',40);
```
unique is missing.

Daniel-Cretu replied:
Added unique to name column by adding an alter migration.

egorbwork replied:
let it be in this way here

egorbwork commented on database/migrations/2021_12_12_172840_create_orders_table.php:
```diff
@@ -0,0 +1,42 @@
+                ->references('id')->on('statuses')
+                ->cascadeOnUpdate()
+                ->restrictOnDelete();
+            $table->decimal('price',10,2)->nullable(false);
```
by default in laravel everything is not nullable, so please ensure that you have desired result on all columns.

Daniel-Cretu replied:
Removed redundant nullable(false) from price column as well as from other migrations.

egorbwork commented on database/migrations/2021_12_12_174255_create_order_products_table.php:
```diff
@@ -0,0 +1,39 @@
+                ->references('id')->on('orders')
+                ->cascadeOnUpdate()
+                ->cascadeOnDelete();
+            $table->foreignId('product_id')->nullable(true)
```
why it should be nullable?

Daniel-Cretu replied:
To be able to delete a product from database without needing to delete all information related to order_products or order_product_toppings.
Example: if we delete an order_product_id from order_products table because product_id got deleted, order_product_id gets NULL value, although we still have order_product_topping_id related to that order_product_id.
On the other hand if we use restrictOnDelete we won't be able to delete any products thus making too much redundancy in our products table.

egorbwork replied:
In general it is not desired to remove products from the database if they are used in the order. So restrict and not null is better in this context. To avoid listing of this products often are used the flag approach, like active column.

egorbwork commented on database/migrations/2021_12_12_174801_create_order_product_toppings_table.php:
```diff
@@ -0,0 +1,39 @@
+                ->references('id')->on('order_products')
+                ->cascadeOnUpdate()
+                ->cascadeOnDelete();
+            $table->bigInteger('product_id')->unsigned()->nullable(true);
```
why it should be nullable?

Daniel-Cretu replied:
Answer mostly similar to the one for order_products_table.

To be able to delete a product from database without needing to delete all information related to order_products or order_product_toppings.
Example: if we delete an order_product_topping_id from order_product_toppings table because product_id got deleted, order_product_topping_id gets NULL value, but we are still left with an number of toppings ordered for a specific product_id thus means we could focus on adding more toppings for that product in the future.
On the other hand if we use restrictOnDelete we won't be able to delete any products thus making too much redundancy in our products table.

egorbwork replied:
What value do you have for removal of the product from database. So you will know that for 400 euro in order were 100 pizza and 50 additional meat for removed Cezar salade?

egorbwork commented:
Also is missing alter table for column change.

Daniel-Cretu commented:
> Also is missing alter table for column change.

Added two alter migrations that changes uniqueness for name columns inside statuses and warnings table.

egorbwork commented:
Column change in migration is still missing.

Daniel-Cretu commented:
> Column change in migration is still missing.

Added column datatype change migration. Removed nullable from order_products and order_product_toppings and set restrictOnDelete

egorbwork reviewed (approved):
Let's give it a try.
